client/src/components/ProductDetails/ProductGallery/Carousel.jsx

```diff
@@ -1,38 +1,150 @@
-import React from 'react';
+import React, { useEffect, useState, useCallback, useRef } from 'react';
+
+const ZOOM_SCALE = 2.5; 
+
+const Carousel = ({
+  photoIndex,
+  photos,
+  setPhotoIndex,
+  styleName,
+  expanded,
+  setExpanded,
+  zoomed,
+  setZoomed,
+  containerRef
+}) => {
+  const imgRef = useRef(null);
+  const [translateX, setTranslateX] = useState(0);
+  const [translateY, setTranslateY] = useState(0);
+
+  const calculateTransformPosition = useCallback((clickEvent) => {
+    const containerBox = containerRef?.current?.getBoundingClientRect();
+    const {naturalWidth, naturalHeight} = imgRef?.current || {};
+    const { top, left, width, height } = containerBox || {};
+
+    /* 
+      The zoomed image height at most is the height of the container * ZOOM_SCALE, otherwise
+      if the image does not take full height of the container (in the case that it 
+      is constrained by its width), it is the its scaled height * ZOOM_SALE 
+    */
```

Comment:
these comments also made it helpful to parse through what your code was intending to do.